File: baekjoon/src/dynamicProgramming/B_1912.java
package dynamicProgramming;

import java.util.*;
import java.io.*;

public class B_1912 {
	static int[] arr;
	static Integer[] dp;

	public static void main(String[] args) throws IOException {
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		
		int N = Integer.parseInt(br.readLine());
		arr = new int[N];
		dp = new Integer[N];
		
		StringTokenizer st = new StringTokenizer(br.readLine());
		for(int i=0;i<N;i++) 
			arr[i] = Integer.parseInt(st.nextToken());
		
		dp[0] = arr[0];
		prefixSum(N - 1);
		
		int max = Integer.MIN_VALUE;
		for(int i=0;i<N;i++)
			max = Math.max(max, dp[i]);
		
		System.out.println(max);
	}

	static int prefixSum(int n) {
		if(dp[n] == null) 
			return dp[n] = Math.max(arr[n] + prefixSum(n - 1), arr[n]);
		
		return dp[n];
	}
}

File: src/kotlin/dp/B_1912.kt
package kotlin.dp

import kotlin.math.max

fun main() = with(System.`in`.bufferedReader()) {
    val n = readLine().toInt()
    val arr = readLine().split(" ").map { it.toInt() }.toIntArray()

    val dp = IntArray(n) { Int.MIN_VALUE }
    dp[0] = arr[0]
    var answer = arr[0]

    for (idx in 1 until n) {
        dp[idx] = max(arr[idx], dp[idx - 1] + arr[idx])
        answer = max(answer, dp[idx])
    }

    print(answer)
}